File: cpp/library_cli.cpp
#include <iostream>
#include <fstream>
#include <filesystem>
#include <string>
#include <vector>
#include <map>
#include <algorithm>
#include <iomanip>
#include <nlohmann/json.hpp>
#include <cxxopts.hpp>
#include <cstdlib>

using json = nlohmann::json;
namespace fs = std::filesystem;

class Library {
private:
    fs::path data_file;
    
    json load_data() {
        if (fs::exists(data_file)) {
            std::ifstream file(data_file);
            json data;
            file >> data;
            return data;
        }
        return json{{"books", json::array()}, {"borrowed", json::object()}};
    }
    
    void save_data(const json& data) {
        std::ofstream file(data_file);
        file << data.dump(2);
    }
    
public:
    Library() {
        const char* home = std::getenv("HOME");
        if (home) {
            data_file = fs::path(home) / ".library_data.json";
        } else {
            data_file = ".library_data.json";
        }
    }
    
    int add_book(const std::string& title, const std::string& author, 
                 const std::string& isbn = "") {
        json data = load_data();
        json book;
        book["id"] = data["books"].size() + 1;
        book["title"] = title;
        book["author"] = author;
        book["isbn"] = isbn.empty() ? 
            "ISBN-" + std::string(4 - std::to_string(data["books"].size() + 1).length(), '0') + 
            std::to_string(data["books"].size() + 1) : isbn;
        book["available"] = true;
        
        data["books"].push_back(book);
        save_data(data);
        
        std::cout << "Added book: " << title << " by " << author 
                  << " (ID: " << book["id"] << ")" << std::endl;
        return book["id"];
    }
    
    void list_books(bool available_only = false) {
        json data = load_data();
        std::vector<json> books = data["books"];
        
        if (available_only) {
            books.erase(
                std::remove_if(books.begin(), books.end(),
                    [](const json& b) { return !b["available"].get<bool>(); }),
                books.end()
            );
        }
        
        if (books.empty()) {
            std::cout << "No books found." << std::endl;
            return;
        }
        
        std::cout << "\n" << std::left << std::setw(5) << "ID" 
                  << std::setw(30) << "Title" 
                  << std::setw(25) << "Author" 
                  << std::setw(15) << "ISBN" 
                  << std::setw(10) << "Status" << std::endl;
        std::cout << std::string(85, '-') << std::endl;
        
        for (const auto& book : books) {
            std::string status = book["available"].get<bool>() ? "Available" : "Borrowed";
            std::cout << std::left << std::setw(5) << book["id"].get<int>()
                      << std::setw(30) << book["title"].get<std::string>()
                      << std::setw(25) << book["author"].get<std::string>()
                      << std::setw(15) << book["isbn"].get<std::string>()
                      << std::setw(10) << status << std::endl;
        }
    }
    
    void borrow_book(int book_id, const std::string& borrower) {
        json data = load_data();
        json* book = nullptr;
        
        for (auto& b : data["books"]) {
            if (b["id"].get<int>() == book_id) {
                book = &b;
                break;
            }
        }
        
        if (!book) {
            std::cout << "Error: Book with ID " << book_id << " not found." << std::endl;
            return;
        }
        
        if (!(*book)["available"].get<bool>()) {
            std::cout << "Error: Book '" << (*book)["title"].get<std::string>() 
                      << "' is already borrowed." << std::endl;
            return;
        }
        
        (*book)["available"] = false;
        data["borrowed"][std::to_string(book_id)] = borrower;
        save_data(data);
        
        std::cout << "'" << (*book)["title"].get<std::string>() 
                  << "' borrowed by " << borrower << std::endl;
    }
    
    void return_book(int book_id) {
        json data = load_data();
        json* book = nullptr;
        
        for (auto& b : data["books"]) {
            if (b["id"].get<int>() == book_id) {
                book = &b;
                break;
            }
        }
        
        if (!book) {
            std::cout << "Error: Book with ID " << book_id << " not found." << std::endl;
            return;
        }
        
        if ((*book)["available"].get<bool>()) {
            std::cout << "Error: Book '" << (*book)["title"].get<std::string>() 
                      << "' is already available." << std::endl;
            return;
        }
        
        std::string borrower = data["borrowed"].contains(std::to_string(book_id)) ?
            data["borrowed"][std::to_string(book_id)].get<std::string>() : "Unknown";
        
        data["borrowed"].erase(std::to_string(book_id));
        (*book)["available"] = true;
        save_data(data);
        
        std::cout << "'" << (*book)["title"].get<std::string>() 
                  << "' returned by " << borrower << std::endl;
    }
    
    void remove_book(int book_id) {
        json data = load_data();
        json* book = nullptr;
        size_t index = 0;
        
        for (size_t i = 0; i < data["books"].size(); ++i) {
            if (data["books"][i]["id"].get<int>() == book_id) {
                book = &data["books"][i];
                index = i;
                break;
            }
        }
        
        if (!book) {
            std::cout << "Error: Book with ID " << book_id << " not found." << std::endl;
            return;
        }
        
        if (!(*book)["available"].get<bool>()) {
            std::cout << "Error: Cannot remove '" << (*book)["title"].get<std::string>() 
                      << "' - it is currently borrowed." << std::endl;
            return;
        }
        
        data["books"].erase(data["books"].begin() + index);
        save_data(data);
        
        std::cout << "Removed book: '" << (*book)["title"].get<std::string>() << "'" << std::endl;
    }
    
    void search_books(const std::string& query) {
        json data = load_data();
        std::string query_lower = query;
        std::transform(query_lower.begin(), query_lower.end(), query_lower.begin(), ::tolower);
        
        std::vector<json> matches;
        for (const auto& book : data["books"]) {
            std::string title = book["title"].get<std::string>();
            std::string author = book["author"].get<std::string>();
            std::transform(title.begin(), title.end(), title.begin(), ::tolower);
            std::transform(author.begin(), author.end(), author.begin(), ::tolower);
            
            if (title.find(query_lower) != std::string::npos || 
                author.find(query_lower) != std::string::npos) {
                matches.push_back(book);
            }
        }
        
        if (matches.empty()) {
            std::cout << "No books found matching '" << query << "'" << std::endl;
            return;
        }
        
        std::cout << "\nFound " << matches.size() << " book(s) matching '" << query << "':" << std::endl;
        std::cout << std::left << std::setw(5) << "ID" 
                  << std::setw(30) << "Title" 
                  << std::setw(25) << "Author" 
                  << std::setw(10) << "Status" << std::endl;
        std::cout << std::string(70, '-') << std::endl;
        
        for (const auto& book : matches) {
            std::string status = book["available"].get<bool>() ? "Available" : "Borrowed";
            std::cout << std::left << std::setw(5) << book["id"].get<int>()
                      << std::setw(30) << book["title"].get<std::string>()
                      << std::setw(25) << book["author"].get<std::string>()
                      << std::setw(10) << status << std::endl;
        }
    }
};

int main(int argc, char* argv[]) {
    Library library;
    
    cxxopts::Options options("library", "Library Management System CLI");
    
    options.add_options()
        ("a,add", "Add a new book", cxxopts::value<std::vector<std::string>>())
        ("l,list", "List all books")
        ("available", "Show only available books")
        ("b,borrow", "Borrow a book", cxxopts::value<std::vector<std::string>>())
        ("r,return", "Return a borrowed book", cxxopts::value<int>())
        ("remove", "Remove a book", cxxopts::value<int>())
        ("s,search", "Search for books", cxxopts::value<std::string>())
        ("h,help", "Print help");
    
    try {
        auto result = options.parse(argc, argv);
        
        if (result.count("help") || argc == 1) {
            std::cout << options.help() << std::endl;
            return 0;
        }
        
        if (result.count("add")) {
            auto args = result["add"].as<std::vector<std::string>>();
            if (args.size() >= 2) {
                std::string isbn = args.size() > 2 ? args[2] : "";
                library.add_book(args[0], args[1], isbn);
            } else {
                std::cout << "Error: add requires title and author" << std::endl;
            }
        } else if (result.count("list")) {
            library.list_books(result.count("available") > 0);
        } else if (result.count("borrow")) {
            auto args = result["borrow"].as<std::vector<std::string>>();
            if (args.size() >= 2) {
                library.borrow_book(std::stoi(args[0]), args[1]);
            } else {
                std::cout << "Error: borrow requires book_id and borrower" << std::endl;
            }
        } else if (result.count("return")) {
            library.return_book(result["return"].as<int>());
        } else if (result.count("remove")) {
            library.remove_book(result["remove"].as<int>());
        } else if (result.count("search")) {
            library.search_books(result["search"].as<std::string>());
        }
        
    } catch (const cxxopts::exceptions::exception& e) {
        std::cerr << "Error parsing options: " << e.what() << std::endl;
        return 1;
    }
    
    return 0;
}

File: python/library_cli.py
#!/usr/bin/env python3
"""
Minimal Library Management System CLI using argparse.
Demonstrates CRUD operations for books with command-line interface.
"""
import argparse
import json
import os
from pathlib import Path

DATA_FILE = Path.home() / '.library_data.json'

def load_data():
    """Load library data from JSON file."""
    if DATA_FILE.exists():
        with open(DATA_FILE, 'r') as f:
            return json.load(f)
    return {'books': [], 'borrowed': {}}

def save_data(data):
    """Save library data to JSON file."""
    with open(DATA_FILE, 'w') as f:
        json.dump(data, f, indent=2)

def add_book(title, author, isbn=None):
    """Add a new book to the library."""
    data = load_data()
    book = {
        'id': len(data['books']) + 1,
        'title': title,
        'author': author,
        'isbn': isbn or f'ISBN-{len(data["books"]) + 1:04d}',
        'available': True
    }
    data['books'].append(book)
    save_data(data)
    print(f"Added book: {title} by {author} (ID: {book['id']})")
    return book['id']

def list_books(available_only=False):
    """List all books in the library."""
    data = load_data()
    books = data['books']
    
    if available_only:
        books = [b for b in books if b['available']]
    
    if not books:
        print("No books found.")
        return
    
    print(f"\n{'ID':<5} {'Title':<30} {'Author':<25} {'ISBN':<15} {'Status':<10}")
    print("-" * 85)
    for book in books:
        status = "Available" if book['available'] else "Borrowed"
        print(f"{book['id']:<5} {book['title']:<30} {book['author']:<25} "
              f"{book['isbn']:<15} {status:<10}")

def borrow_book(book_id, borrower):
    """Borrow a book from the library."""
    data = load_data()
    book = next((b for b in data['books'] if b['id'] == book_id), None)
    
    if not book:
        print(f"Error: Book with ID {book_id} not found.")
        return
    
    if not book['available']:
        print(f"Error: Book '{book['title']}' is already borrowed.")
        return
    
    book['available'] = False
    data['borrowed'][str(book_id)] = borrower
    save_data(data)
    print(f"'{book['title']}' borrowed by {borrower}")

def return_book(book_id):
    """Return a borrowed book to the library."""
    data = load_data()
    book = next((b for b in data['books'] if b['id'] == book_id), None)
    
    if not book:
        print(f"Error: Book with ID {book_id} not found.")
        return
    
    if book['available']:
        print(f"Error: Book '{book['title']}' is already available.")
        return
    
    borrower = data['borrowed'].pop(str(book_id), 'Unknown')
    book['available'] = True
    save_data(data)
    print(f"'{book['title']}' returned by {borrower}")

def remove_book(book_id):
    """Remove a book from the library."""
    data = load_data()
    book = next((b for b in data['books'] if b['id'] == book_id), None)
    
    if not book:
        print(f"Error: Book with ID {book_id} not found.")
        return
    
    if not book['available']:
        print(f"Error: Cannot remove '{book['title']}' - it is currently borrowed.")
        return
    
    data['books'] = [b for b in data['books'] if b['id'] != book_id]
    save_data(data)
    print(f"Removed book: '{book['title']}'")

def search_books(query):
    """Search for books by title or author."""
    data = load_data()
    query_lower = query.lower()
    matches = [
        b for b in data['books']
        if query_lower in b['title'].lower() or query_lower in b['author'].lower()
    ]
    
    if not matches:
        print(f"No books found matching '{query}'")
        return
    
    print(f"\nFound {len(matches)} book(s) matching '{query}':")
    print(f"{'ID':<5} {'Title':<30} {'Author':<25} {'Status':<10}")
    print("-" * 70)
    for book in matches:
        status = "Available" if book['available'] else "Borrowed"
        print(f"{book['id']:<5} {book['title']:<30} {book['author']:<25} {status:<10}")

def main():
    parser = argparse.ArgumentParser(
        description='Library Management System CLI',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  %(prog)s add "The Great Gatsby" "F. Scott Fitzgerald"
  %(prog)s list
  %(prog)s borrow 1 "John Doe"
  %(prog)s return 1
  %(prog)s search "Gatsby"
  %(prog)s remove 1
        """
    )
    
    subparsers = parser.add_subparsers(dest='command', help='Available commands')
    
    # Add command
    add_parser = subparsers.add_parser('add', help='Add a new book')
    add_parser.add_argument('title', help='Book title')
    add_parser.add_argument('author', help='Book author')
    add_parser.add_argument('--isbn', help='ISBN number (optional)')
    
    # List command
    list_parser = subparsers.add_parser('list', help='List all books')
    list_parser.add_argument('--available', action='store_true', 
                            help='Show only available books')
    
    # Borrow command
    borrow_parser = subparsers.add_parser('borrow', help='Borrow a book')
    borrow_parser.add_argument('book_id', type=int, help='Book ID')
    borrow_parser.add_argument('borrower', help='Borrower name')
    
    # Return command
    return_parser = subparsers.add_parser('return', help='Return a borrowed book')
    return_parser.add_argument('book_id', type=int, help='Book ID')
    
    # Remove command
    remove_parser = subparsers.add_parser('remove', help='Remove a book')
    remove_parser.add_argument('book_id', type=int, help='Book ID')
    
    # Search command
    search_parser = subparsers.add_parser('search', help='Search for books')
    search_parser.add_argument('query', help='Search term (title or author)')
    
    args = parser.parse_args()
    
    if not args.command:
        parser.print_help()
        return
    
    if args.command == 'add':
        add_book(args.title, args.author, args.isbn)
    elif args.command == 'list':
        list_books(available_only=args.available)
    elif args.command == 'borrow':
        borrow_book(args.book_id, args.borrower)
    elif args.command == 'return':
        return_book(args.book_id)
    elif args.command == 'remove':
        remove_book(args.book_id)
    elif args.command == 'search':
        search_books(args.query)

if __name__ == '__main__':
    main()
